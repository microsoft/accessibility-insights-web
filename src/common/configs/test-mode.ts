// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
export enum TestMode {
    Adhoc = 'adhoc',
    Assessments = 'assessments',
    QuickAssess = 'quickAssess',
}
