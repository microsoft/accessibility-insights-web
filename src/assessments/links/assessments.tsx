// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { VisualizationType } from 'common/types/visualization-type';
import { test as content } from 'content/test';
import * as React from 'react';
import { AssessmentBuilder } from '../assessment-builder';
import * as Markup from '../markup';
import { LabelInName } from './test-steps/label-in-name';
import { LinkFunction } from './test-steps/link-function';
import { LinkPurpose } from './test-steps/link-purpose';

const key = 'linksAssessment';
const title = 'Links';
const { guidance } = content.links;

const LinksAssessmentGettingStarted: JSX.Element = (
    <React.Fragment>
        <p>
            A link is a specific type of widget (interactive interface component) that navigates the
            user to new content, either in the current page or in a new page. A link is typically
            implemented as an HTML <Markup.Term>a</Markup.Term> (anchor) element with an{' '}
            <Markup.Term>href</Markup.Term> value.
        </p>
        However, <Markup.Term>a</Markup.Term> elements are frequently scripted to function as
        buttons. The difference between a link and a button:
        <ul>
            <li>
                A <Markup.Term>link</Markup.Term> <Markup.Emphasis>navigates</Markup.Emphasis> to
                new content.
            </li>
            <li>
                A <Markup.Term>button</Markup.Term> <Markup.Emphasis>activates</Markup.Emphasis>{' '}
                functionality (e.g., shows or hides content, toggles a state on or off).
            </li>
        </ul>
        For example, a navigation bar is a list of links, while a menu is a collection of buttons.
        <p>
            An <Markup.Term>a</Markup.Term> element that functions as a button or other custom
            widget must have the appropriate widget role.
        </p>
    </React.Fragment>
);

export const LinksAssessment = AssessmentBuilder.Assisted({
    key,
    title: title,
    gettingStarted: LinksAssessmentGettingStarted,
    guidance,
    visualizationType: VisualizationType.LinksAssessment,
    requirements: [LinkFunction, LinkPurpose, LabelInName],
    storeDataKey: 'linksAssessment',
});
