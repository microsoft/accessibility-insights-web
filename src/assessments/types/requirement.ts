// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { IColumn } from '@fluentui/react';
import { UniquelyIdentifiableInstances } from 'background/instance-identifier-generator';
import { HyperlinkDefinition } from 'common/types/hyperlink-definition';
import { ManualTestStatus } from 'common/types/manual-test-status';
import {
    AssessmentNavState,
    GeneratedAssessmentInstance,
    InstanceIdToInstanceDataMap,
} from 'common/types/store-data/assessment-result-data';
import { FeatureFlagStoreData } from 'common/types/store-data/feature-flag-store-data';
import { DetailsViewActionMessageCreator } from 'DetailsView/actions/details-view-action-message-creator';
import { Analyzer } from 'injected/analyzers/analyzer';
import { AnalyzerProvider } from 'injected/analyzers/analyzer-provider';
import { DecoratedAxeNodeResult } from 'injected/scanner-utils';
import { VisualizationInstanceProcessorCallback } from 'injected/visualization-instance-processor';
import { Drawer } from 'injected/visualization/drawer';
import { DrawerProvider } from 'injected/visualization/drawer-provider';
import { DictionaryStringTo } from 'types/common-types';
import { ContentPageComponent } from 'views/content/content-page';
import { IGetMessageGenerator } from '../assessment-default-message-generator';
import { InstanceTableColumn, InstanceTableHeaderType } from './instance-table-data';
import { ReportInstanceFields } from './report-instance-field';

export interface Requirement {
    key: string;
    description: JSX.Element;
    order?: number;
    name: string;
    renderStaticContent?: () => JSX.Element;
    howToTest: JSX.Element;
    addFailureInstruction?: string;
    infoAndExamples?: ContentPageComponent;
    isManual: boolean;
    // This is for semi-manual cases where we can't present a list of instances like an assisted
    // test would, but can infer a PASS or FAIL state. If not specified, acts like () => UNKNOWN.
    getInitialManualTestStatus?: (instances: InstanceIdToInstanceDataMap) => ManualTestStatus;
    guidanceLinks: HyperlinkDefinition[];
    columnsConfig?: InstanceTableColumn[];
    getAnalyzer?: (provider: AnalyzerProvider) => Analyzer;
    getVisualHelperToggle?: (props: VisualHelperToggleConfig) => JSX.Element;
    // Any results for which this returns false will be omitted from visual helper displays, but still
    // present for the purposes of instance lists or getInitialManualTestStatus. By default, all
    // results support visualization.
    isVisualizationSupportedForResult?: (result: DecoratedAxeNodeResult) => boolean;
    visualizationInstanceProcessor?: VisualizationInstanceProcessorCallback;
    getDrawer?: (provider: DrawerProvider, featureFlagStoreData?: FeatureFlagStoreData) => Drawer;
    getNotificationMessage?: (selectorMap: DictionaryStringTo<any>) => string;
    doNotScanByDefault?: boolean;
    switchToTargetTabOnScan?: boolean;
    generateInstanceIdentifier?: (instance: UniquelyIdentifiableInstances) => string;
    reportInstanceFields?: ReportInstanceFields;
    renderReportDescription?: () => JSX.Element;
    getInstanceStatus?: (result: DecoratedAxeNodeResult) => ManualTestStatus;
    getInstanceStatusColumns?: () => Readonly<IColumn>[];
    getDefaultMessage?: IGetMessageGenerator;
    instanceTableHeaderType?: InstanceTableHeaderType;
}

export type VisualHelperToggleConfigDeps = {
    detailsViewActionMessageCreator: DetailsViewActionMessageCreator;
};

export interface VisualHelperToggleConfig {
    deps: VisualHelperToggleConfigDeps;
    assessmentNavState: AssessmentNavState;
    instancesMap: DictionaryStringTo<GeneratedAssessmentInstance>;
    isStepEnabled: boolean;
    isStepScanned: boolean;
}
