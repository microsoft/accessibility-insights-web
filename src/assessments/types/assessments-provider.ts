// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { VisualizationType } from 'common/types/visualization-type';
import { DictionaryStringTo } from 'types/common-types';
import { Assessment } from './iassessment';
import { Requirement } from './requirement';

export interface AssessmentsProvider {
    all(): ReadonlyArray<Readonly<Assessment>>;
    isValidType(visualizationType: VisualizationType): boolean;
    forType(visualizationType: VisualizationType): Readonly<Assessment>;
    isValidKey(key: string): boolean;
    forKey(key: string): Readonly<Assessment>;
    forRequirementKey(key: string): Readonly<Assessment>;
    getStep(visualizationType: VisualizationType, key: string): Readonly<Requirement>;
    getStepMap(
        visualizationType: VisualizationType,
    ): Readonly<DictionaryStringTo<Readonly<Requirement>>>;
}
