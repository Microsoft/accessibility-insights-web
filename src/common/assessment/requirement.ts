// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { keys, sortBy } from 'lodash';

import { IAssessmentsProvider } from '../../assessments/types/iassessments-provider';
import { TestStep } from '../../assessments/types/test-step';
import { ManualTestStatusData, TestStepData } from '../types/manual-test-status';
import { VisualizationType } from '../types/visualization-type';

export type RequirementDefinition = TestStep;
export type RequirementData = TestStepData;

export type RequirementResult = {
    definition: RequirementDefinition;
    data: RequirementData;
};
export type RequirementOrdering = RequirementOrderPart | RequirementOrderPart[];
export type RequirementOrderPart = (result: RequirementResult) => string | number;

export function getRequirementsResults(
    provider: IAssessmentsProvider,
    type: VisualizationType,
    stepStatus: ManualTestStatusData,
): RequirementResult[] {
    const test = provider.forType(type);

    function result(key): RequirementResult {
        const definition = provider.getStep(type, key);
        const data = stepStatus[key];
        return { definition, data };
    }

    const unsorted = keys(stepStatus).map(result);

    return sortBy(unsorted, test.requirementOrder);
}
