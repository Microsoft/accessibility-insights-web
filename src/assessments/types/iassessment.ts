// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { RequirementOrdering } from '../../common/assessment/requirement';
import { IAssesssmentVisualizationConfiguration } from '../../common/configs/visualization-configuration-factory';
import { AnyExtension } from '../../common/extensibility/extension-point';
import { IAssessmentData } from '../../common/types/store-data/iassessment-result-data';
import { VisualizationType } from '../../common/types/visualization-type';
import { ContentPageComponent } from '../../views/content/content-page';
import { TestStep } from './test-step';

interface BaseAssessment {
    key: string;
    type: VisualizationType;
    title: string;
    gettingStarted: JSX.Element;
    guidance?: ContentPageComponent;
    steps: TestStep[];
    featureFlag?: { required?: string[] };
    executeAssessmentScanPolicy?: (scheduleScan: (step: string) => void, data: IAssessmentData) => void;
    requirementOrder?: RequirementOrdering;
    extensions?: AnyExtension[];
}

export interface ManualAssessment extends BaseAssessment {}

export interface AssistedAssessment extends BaseAssessment {
    storeDataKey: string;
    visualizationConfiguration?: Partial<IAssesssmentVisualizationConfiguration>;
}

export interface Assessment extends BaseAssessment {
    getVisualizationConfiguration: () => IAssesssmentVisualizationConfiguration;
    requirementOrder: RequirementOrdering;
}
