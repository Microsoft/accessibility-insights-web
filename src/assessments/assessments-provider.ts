// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { VisualizationType } from '../common/types/visualization-type';
import { Assessment } from './types/iassessment';
import { IAssessmentsProvider } from './types/iassessments-provider';
import { TestStep } from './types/test-step';

export class AssessmentsProvider implements IAssessmentsProvider {
    private assessments: Assessment[];
    public static Create(assessments: Assessment[]): IAssessmentsProvider {
        const provider = new AssessmentsProvider();
        provider.assessments = assessments.slice();
        return provider;
    }

    public all(): Assessment[] {
        return this.assessments.slice();
    }

    public forType(type: VisualizationType): Assessment {
        return this.all().find(a => a.type === type);
    }

    public isValidType(type: VisualizationType): boolean {
        return this.forType(type) != null;
    }

    public forKey(key: string): Assessment {
        return this.all().find(a => a.key === key);
    }

    public isValidKey(key: string): boolean {
        return this.forKey(key) != null;
    }

    public getStep(type: VisualizationType, key: string): TestStep {
        const assessment = this.forType(type);
        if (!assessment) {
            return null;
        }
        const steps = assessment.steps;
        const index = steps.findIndex(s => s.key === key);
        if (index === -1) {
            return null;
        }
        return { ...steps[index], order: index + 1 };
    }

    public getStepMap(type: VisualizationType): IDictionaryStringTo<TestStep> {
        const assessment = this.forType(type);
        if (!assessment) {
            return null;
        }
        const steps = assessment.steps;

        let index = 1;
        return steps.reduce((map, step) => {
            map[step.key] = { ...step, order: index++ };
            return map;
        }, {});
    }
}
