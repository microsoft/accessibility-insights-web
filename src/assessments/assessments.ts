// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { AdaptableContentAssessment } from './adaptable-content/assessment';
import { AssessmentsProviderImpl } from './assessments-provider';
import { AudioVideoOnlyAssessment } from './audio-video-only/assessment';
import { AutomatedChecks } from './automated-checks/assessment';
import { CognitiveAssessment } from './cognitive/assessment';
import { ColorSensoryAssessment } from './color/assessment';
import { ContrastAssessment } from './contrast/assessment';
import { CustomWidgets } from './custom-widgets/assessment';
import { ErrorsAssessment } from './errors/assessment';
import { HeadingsAssessment } from './headings/assessment';
import { ImagesAssessment } from './images/assessment';
import { KeyboardInteraction } from './keyboard-interaction/assessment';
import { LandmarksAssessment } from './landmarks/assessment';
import { LanguageAssessment } from './language/assessment';
import { LinksAssessment } from './links/assessments';
import { LiveMultimediaAssessment } from './live-multimedia/assessment';
import { NativeWidgetsAssessment } from './native-widgets/assessment';
import { PageAssessment } from './page/assessment';
import { ParsingAssessment } from './parsing/assessment';
import { PointerMotionAssessment } from './pointer-motion/assessment';
import { PrerecordedMultimediaAssessment } from './prerecorded-multimedia/assessment';
import { RepetitiveContentAssessment } from './repetitive-content/assessment';
import { SemanticsAssessment } from './semantics/assessment';
import { SequenceAssessment } from './sequence/assessment';
import { TimedEventsAssessment } from './timed-events/assessment';
import { AssessmentsProvider } from './types/assessments-provider';
import { VisibleFocusOrderAssessment } from './visible-focus-order/assessment';

export const Assessments: AssessmentsProvider = AssessmentsProviderImpl.Create([
    AutomatedChecks,
    KeyboardInteraction,
    VisibleFocusOrderAssessment,
    LandmarksAssessment,
    HeadingsAssessment,
    RepetitiveContentAssessment,
    LinksAssessment,
    NativeWidgetsAssessment,
    CustomWidgets,
    TimedEventsAssessment,
    ErrorsAssessment,
    PageAssessment,
    ParsingAssessment,
    ImagesAssessment,
    LanguageAssessment,
    ColorSensoryAssessment,
    AdaptableContentAssessment,
    AudioVideoOnlyAssessment,
    PrerecordedMultimediaAssessment,
    LiveMultimediaAssessment,
    SequenceAssessment,
    SemanticsAssessment,
    PointerMotionAssessment,
    ContrastAssessment,
    CognitiveAssessment,
]);
