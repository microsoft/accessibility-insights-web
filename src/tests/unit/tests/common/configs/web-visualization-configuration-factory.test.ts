// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { Assessments } from 'assessments/assessments';
import { AssessmentsProvider } from 'assessments/types/assessments-provider';
import { Assessment } from 'assessments/types/iassessment';
import { Requirement } from 'assessments/types/requirement';
import { TestMode } from 'common/configs/test-mode';
import { VisualizationConfiguration } from 'common/configs/visualization-configuration';
import { WebVisualizationConfigurationFactory } from 'common/configs/web-visualization-configuration-factory';
import { forOwn } from 'lodash';
import { IMock, It, Mock, Times } from 'typemoq';
import { DictionaryNumberTo } from 'types/common-types';
import {
    AssessmentData,
    AssessmentStoreData,
    PersistedTabInfo,
} from '../../../../../common/types/store-data/assessment-result-data';
import {
    ScanData,
    TestsEnabledState,
    TestsScanData,
} from '../../../../../common/types/store-data/visualization-store-data';
import { VisualizationType } from '../../../../../common/types/visualization-type';

describe('WebVisualizationConfigurationFactory', () => {
    let testObject: WebVisualizationConfigurationFactory;

    let mediumPassProviderMock: IMock<AssessmentsProvider>;

    beforeEach(() => {
        mediumPassProviderMock = Mock.ofType<AssessmentsProvider>();
        mediumPassProviderMock.setup(m => m.isValidType(It.isAny())).returns(() => false);
        testObject = new WebVisualizationConfigurationFactory(
            Assessments,
            mediumPassProviderMock.object,
        );
    });

    test('get config for unsupported type', () => {
        const invalidType = -1 as VisualizationType;
        const action = () => {
            testObject.getConfiguration(invalidType);
        };

        expect(action).toThrowError(`Unsupported type: ${invalidType}`);
    });

    test('getStoreData for color', () => {
        const visualizationType = VisualizationType.Color;
        const getExpectedData: (data: TestsEnabledState) => ScanData = data => data.adhoc.color;

        testGetStoreData(visualizationType, getExpectedData);
    });

    test('getStoreData for headings', () => {
        const visualizationType = VisualizationType.Headings;
        const getExpectedData: (data: TestsEnabledState) => ScanData = data => data.adhoc.headings;

        testGetStoreData(visualizationType, getExpectedData);
    });

    test('getStoreData for headingsAssessment', () => {
        const visualizationType = VisualizationType.HeadingsAssessment;
        const getExpectedData: (data: TestsEnabledState) => ScanData = data =>
            data.assessments.headingsAssessment;

        testGetStoreData(visualizationType, getExpectedData);
    });

    test('setAssessmentData for headingsAssessment', () => {
        const visualizationType = VisualizationType.HeadingsAssessment;
        const testData: AssessmentStoreData = {
            persistedTabInfo: {} as PersistedTabInfo,
            assessments: {
                headings: {
                    fullAxeResultsMap: null,
                    generatedAssessmentInstancesMap: null,
                } as AssessmentData,
            },
            assessmentNavState: null,
            resultDescription: '',
        };
        const selectorMap = {
            selector: {},
        };
        const instanceMap = {
            selector: {},
        };
        const configuration = testObject.getConfiguration(visualizationType);
        configuration.setAssessmentData(testData, selectorMap, instanceMap);
        expect(testData.assessments.headings.fullAxeResultsMap).toEqual(selectorMap);
    });

    test('getStoreData for issues', () => {
        const visualizationType = VisualizationType.Issues;
        const getExpectedData: (data: TestsEnabledState) => ScanData = data => data.adhoc.issues;

        testGetStoreData(visualizationType, getExpectedData);
    });

    test('getStoreData for landmarks', () => {
        const visualizationType = VisualizationType.Landmarks;
        const getExpectedData: (data: TestsEnabledState) => ScanData = data => data.adhoc.landmarks;

        testGetStoreData(visualizationType, getExpectedData);
    });

    test('getStoreData for tabStops', () => {
        const visualizationType = VisualizationType.TabStops;
        const getExpectedData: (data: TestsEnabledState) => ScanData = data => data.adhoc.tabStops;

        testGetStoreData(visualizationType, getExpectedData);
    });

    test('displayableData for color', () => {
        testDisplayableData(VisualizationType.Color);
    });

    test('displayableData for headings', () => {
        testDisplayableData(VisualizationType.Headings);
    });

    test('displayableData for headings', () => {
        testDisplayableData(VisualizationType.HeadingsAssessment);
    });

    test('displayableData for issues', () => {
        testDisplayableData(VisualizationType.Issues);
    });

    test('displayableData for landmarks', () => {
        testDisplayableData(VisualizationType.Landmarks);
    });

    test('displayableData for tabStops', () => {
        testDisplayableData(VisualizationType.TabStops);
    });

    test('getChromeCommandToVisualizationTypeMap', () => {
        const result = testObject.getChromeCommandToVisualizationTypeMap();
        expect(result).toMatchSnapshot();
    });

    test('getChromeCommandToVisualizationTypeMap where a config does not have a chrome command', () => {
        const configurationByTypeStub = {
            [VisualizationType.Color]: {} as VisualizationConfiguration,
        } as DictionaryNumberTo<VisualizationConfiguration>;
        (testObject as any).configurationByType = {
            ...(testObject as any).configurationByType,
            ...configurationByTypeStub,
        };
        const result = testObject.getChromeCommandToVisualizationTypeMap();
        expect(result).toMatchSnapshot();
    });

    test('forEachConfig', () => {
        const assessmentProviderMock = Mock.ofType<AssessmentsProvider>();
        const mediumPassProviderMock = Mock.ofType<AssessmentsProvider>();
        const callbackMock =
            Mock.ofType<
                (
                    config: VisualizationConfiguration,
                    type: VisualizationType,
                    requirementConfig?: Requirement,
                ) => void
            >();

        testObject = new WebVisualizationConfigurationFactory(
            assessmentProviderMock.object,
            mediumPassProviderMock.object,
        );
        const assessmentStubs = [getAssessmentStub('a-1', -1), getAssessmentStub('a-2', -2)];
        const mediumPassStubs = [getAssessmentStub('mp-1', -3), getAssessmentStub('mp-2', -4)];

        mediumPassProviderMock.setup(mock => mock.all()).returns(() => mediumPassStubs);
        assessmentProviderMock.setup(mock => mock.all()).returns(() => assessmentStubs);

        testObject.forEachConfig(callbackMock.object);

        verifyEachProviderConfigIsCalled(callbackMock, assessmentStubs, TestMode.Assessments);
        verifyEachProviderConfigIsCalled(callbackMock, mediumPassStubs, TestMode.MediumPass);
        forOwn(
            (testObject as any).configurationByType,
            (config: VisualizationConfiguration, key: string) => {
                const type = Number(key);
                callbackMock.verify(m => m(config, type), Times.once());
            },
        );
    });

    function getAssessmentStub(title: string, visualizationType: VisualizationType): Assessment {
        const requirementsStub = [
            {
                key: 'req-1',
            },
            {
                key: 'req-2',
            },
        ] as Requirement[];
        return {
            title,
            visualizationType,
            requirements: requirementsStub,
            getVisualizationConfiguration: () => {},
        } as Assessment;
    }

    function verifyEachProviderConfigIsCalled(
        callbackMock: IMock<
            (
                config: VisualizationConfiguration,
                type: VisualizationType,
                requirementConfig?: Requirement,
            ) => void
        >,
        assessmentStubs: Readonly<Assessment>[],
        testMode: TestMode,
    ) {
        assessmentStubs.forEach(stub => {
            stub.requirements.forEach(reqStub =>
                callbackMock.verify(
                    m =>
                        m(
                            It.isObjectWith(getAssessmentDefaults(stub.title, testMode)),
                            stub.visualizationType,
                            reqStub,
                        ),
                    Times.once(),
                ),
            );
        });
    }

    test('getConfiguration for mediumPass', () => {
        const type = VisualizationType.HeadingsAssessment;
        const requirementKey = 'some requirement key';
        const visualizationKeyStub = 'some key';
        const assessmentStub = {
            title: 'some title',
            getVisualizationConfiguration: () => ({ key: visualizationKeyStub }),
            requirements: [
                {
                    key: requirementKey,
                },
            ],
        } as Assessment;
        const testData = {
            [TestMode.MediumPass]: {
                [visualizationKeyStub]: {
                    enabled: true,
                },
            } as TestsScanData,
        } as TestsEnabledState;
        const expectedDefaults = getAssessmentDefaults(assessmentStub.title, TestMode.MediumPass);
        const expected = {
            ...assessmentStub.getVisualizationConfiguration(),
            ...expectedDefaults,
        };
        mediumPassProviderMock.reset();
        mediumPassProviderMock.setup(m => m.isValidType(type)).returns(() => true);
        mediumPassProviderMock.setup(m => m.forType(type)).returns(() => assessmentStub);

        const returnedConfiguration = testObject.getConfiguration(type);
        expect(returnedConfiguration).toMatchObject(expected);
        expect(returnedConfiguration.getIdentifier(requirementKey)).toEqual(
            `${TestMode.MediumPass}-${requirementKey}`,
        );
        expect(returnedConfiguration.getStoreData(testData)).toEqual(
            testData.mediumPass[visualizationKeyStub],
        );
    });

    test('getConfigurationByKey for adhoc visualizations', () => {
        const keys = ['color', 'headings', 'landmarks', 'tabStops', 'issues'];

        keys.forEach(key => {
            const configuration = testObject.getConfigurationByKey(key);

            expect(configuration).toBeDefined();
            expect(configuration.key).toBe(key);
        });
    });

    // This is important for any data structure which assumes it's safe to use configIds as keys
    test("manually specified visualizaton keys don't overlap with requirement visualization keys", () => {
        for (const assessment of Assessments.all()) {
            const assessmentConfig = testObject.getConfiguration(assessment.visualizationType);
            for (const requirement of assessment.requirements) {
                const requirementVisualizationKey = assessmentConfig.getIdentifier(requirement.key);
                const correspondingManuallyKeyedConfiguration = testObject.getConfigurationByKey(
                    requirementVisualizationKey,
                );

                expect(correspondingManuallyKeyedConfiguration).toBeUndefined();
            }
        }
    });

    function testDisplayableData(visualizationType: VisualizationType): void {
        const configuration = testObject.getConfiguration(visualizationType);
        const displayableData = configuration.displayableData;

        expect(displayableData).toBeDefined();
    }

    function testGetStoreData(
        visualizationType: VisualizationType,
        getExpectedData: (data: TestsEnabledState) => ScanData,
    ): void {
        const configuration = testObject.getConfiguration(visualizationType);
        const data = {
            [configuration.testMode]: {
                [configuration.key]: {
                    enabled: true,
                },
            },
        } as any;
        const scanData = configuration.getStoreData(data);

        const expected = getExpectedData(data);

        expect(scanData).toEqual(expected);
    }

    function getAssessmentDefaults(
        expectedDisplayableTitle: string,
        testMode: TestMode,
    ): Partial<VisualizationConfiguration> {
        return {
            testMode,
            chromeCommand: null,
            launchPanelDisplayOrder: null,
            adhocToolsPanelDisplayOrder: null,
            displayableData: {
                title: expectedDisplayableTitle,
                noResultsFound: null,
                enableMessage: null,
                toggleLabel: null,
                linkToDetailsViewText: null,
            },
            shouldShowExportReport: null,
        };
    }
});
