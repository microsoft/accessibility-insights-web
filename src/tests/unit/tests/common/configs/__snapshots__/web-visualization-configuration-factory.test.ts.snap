// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WebVisualizationConfigurationFactory getChromeCommandToVisualizationTypeMap 1`] = `
{
  "01_toggle-issues": 1,
  "02_toggle-landmarks": 2,
  "03_toggle-headings": 0,
  "04_toggle-tabStops": 3,
  "05_toggle-color": 4,
  "06_toggle-needsReview": 29,
  "07_toggle-accessibleNames": 30,
}
`;

exports[`WebVisualizationConfigurationFactory getChromeCommandToVisualizationTypeMap where a config does not have a chrome command 1`] = `
{
  "01_toggle-issues": 1,
  "02_toggle-landmarks": 2,
  "03_toggle-headings": 0,
  "04_toggle-tabStops": 3,
  "06_toggle-needsReview": 29,
  "07_toggle-accessibleNames": 30,
}
`;
